momepy/preprocessing.py

```diff
@@ -740,3 +745,368 @@ def _get_extrapolated_line(coords, tolerance, point=False):
+
+    1. All nodes have attributes determining their x and y coordinates;
+    2. All edges have attributes determining their origin, destination, and geometry;
+    3. Graph itself has an attribute determining its crs.
+
+    Parameters
+    ----------
+    graph : Networkx.MultiGraph or Networkx.MultiDiGraph
+
+    tolerance : float
+        distance in network units below which nodes will be consolidated
+
+    rebuild_graph: Boolean
+
+    rebuild_edges_method: string
+        'spider' or 'linear' or 'crow_flies'
+
+    directed: Boolean
+
+    x_col, y_col: string
+        node attribute with the valid coordinate
+
+    edge_from_col, edge_to_col: string
+        edge attribute with the valid origin/destination node id
+
+    Returns
+    ----------
+    Networkx.MultiGraph or Networkx.MultiDiGraph
+
+    """
+
+    # Collect nodes and their data:
+    nodes, nodes_dict = zip(*graph.nodes(data=True))
+    nodes_df = pd.DataFrame(nodes_dict, index=nodes)
+    nodes_gdf = gpd.GeoDataFrame(
+        nodes_df,
+        crs=graph.graph["crs"],
+        geometry=gpd.points_from_xy(nodes_df[x_col], nodes_df[y_col]),
+    )  # improve crs here
+
+    # In case we did not specify directionality, we infer it from the network:
+    if directed is None:
+        directed = True if isinstance(graph, nx.MultiDiGraph) else False
+
+    # Create a graph without the edges above a certain length and clean it from isolated nodes (the unsimplifiable nodes):
+    components_graph = deepcopy(graph)
+    components_graph.remove_edges_from(
+        [
+            edge
+            for edge in graph.edges(keys=True, data=True)
+            if edge[-1]["length"] > tolerance
+        ]
+    )
+    isolated_nodes_list = list(nx.isolates(components_graph))
+    components_graph.remove_nodes_from(isolated_nodes_list)
+
+    # The connected components of this graph are node clusters we must individually simplify. We collect them in a dataframe and
+    #  retrieve node properties (x, y coords mainly) from the original graph.
+    components_dict = dict(
+        enumerate(nx.connected_components(components_graph), start=max(nodes) + 1)
+    )
+    nodes_to_merge_dict = {
+        node: cpt for cpt, nodes in components_dict.items() for node in nodes
+    }
+    nodes_to_merge_df = pd.concat(
+        [
+            pd.DataFrame.from_dict(
+                nodes_to_merge_dict, orient="index", columns=["cluster"]
+            ),
+            nodes_df[[x_col, y_col]],
+        ],
+        axis=1,
+        join="inner",
+    )
+
+    # The two node attributes we need for the clusters are the position of the cluster centroids. Those are obtained by averaging
+    #  the x and y columns. We also add attribtues referring to the original osmids in every cluster:
+    cluster_centroids_df = nodes_to_merge_df.groupby("cluster").mean()
+    cluster_centroids_df["simplified"] = True
+    cluster_centroids_df["original_osmids"] = cluster_centroids_df.index.map(
```

Comment:
This assumes `osmnx`-like graph, doesn't it? At least in the name.

Reply:
good catch. is "original_nodes" a better name?

Reply:
Yeah